### PR TITLE
Add: Probing ViTs

### PR DESCRIPTION
**Pull request by ariG23498**
We (/w @sayakpaul) probe into the representations learned by different vision transformer (ViT) models. Our main goal with this example is to provide insights into what empowers ViTs to learn from image data. In particular, the example discusses
implementations of a few different ViT analysis tools.

We hope this helps the entire TensorFlow community.

[Colab notebook](https://colab.research.google.com/drive/1Y38JFHZpEKv5w9uNg6KbIE-jC2QATdrB?usp=sharing)

[Hugging face organisation](https://huggingface.co/probing-vits)

### CODE REVIEW

File: examples/vision/probing_vits.py [hunk 1]
```diff
@@ -0,0 +1,560 @@
...
+(DINO).
+
+To keep the example concise, we'll not exhaustively pair each model with the analysis
+methods. But we'll provide notes in the respective sections so that you can pick up the
+pieces. 
+
+To run this example on Colab, we need to update the `gdown` library like so:
+
+```shell
+$ pip install -U gdown -q
```
Comment (fchollet) [inline examples/vision/probing_vits.py:43]: No need for the `$`

File: examples/vision/probing_vits.py [hunk 2]
```diff
@@ -0,0 +1,560 @@
...
+* Another repository that you may find interesting in this regard is
+[`vit-explain`](https://github.com/jacobgil/vit-explain).
+
+* One can also plot the attention rollout and attention heat maps with
+custom images using our Hugging Face spaces.
+
+| Attention Heat Maps | Attention Rollout |
+| :--: | :--: |
+| [![Generic badge](https://img.shields.io/badge/🤗%20Spaces-Attention%20Heat%20Maps-black.svg)](https://huggingface.co/spaces/probing-vits/attention-heat-maps) |
+[![Generic badge](https://img.shields.io/badge/🤗%20Spaces-Attention%20Rollout-black.svg)](https://huggingface.co/spaces/probing-vits/attention-rollout) |
```
Comment (fchollet) [inline examples/vision/probing_vits.py:551]: They even have an emoji in the URL!

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! This is another very neat example.

It doesn't leverage a lot of Keras functionality (save for showing how to use TF-Hub with Keras), but nevertheless it should a very nice addition.

I will do a copyedits pass a bit later.

**Comment by sayakpaul** [comment]
> It doesn't leverage a lot of Keras functionality (save for showing how to use TF-Hub with Keras), but nevertheless it should a very nice addition.

I completely agree, @fchollet. 

Our motivation was TF/Keras users should be able to investigate ViT models like they do for CNNs. Since CNNs have their guides when it comes to investigating them in Keras we thought why not do it for ViTs too. 

This is why a journey started. We implemented a few ViT model families in Keras (which we link from the example), ported the pre-trained parameters into them, and verified their results as reported in the respective papers. I think this might be a good showcase of Keras functionalities especially how the models are implemented. Should we call this out from the tutorial?

FYI, we are also working on two more investigation method that would allow readers to know how do the linear projection and positional encodings look like. We will add them in the example as soon as we're done.

File: examples/vision/probing_vits.py (hunk 2 above)
Comment (sayakpaul) [reply examples/vision/probing_vits.py:551]: It was to generate the text on the graphic badge.

**Comment by ariG23498** [comment]
Hey @fchollet 

> FYI, we are also working on two more investigation method that would allow readers to know how do the linear projection and positional encodings look like. We will add them in the example as soon as we're done.

We have completed our investigation and have added it along with our observations. Please feel free to review whenever possible.

File: examples/vision/probing_vits.py [hunk 3]
```diff
@@ -0,0 +1,679 @@
+"""
+Title: Under the Hood of Vision Transformers
```
Comment (fchollet) [inline examples/vision/probing_vits.py:2]: Let's find a better title. "Explainability for Vision Transformers" / "Investigating Vision Transformers representations" / something like that.

File: examples/vision/probing_vits.py [hunk 4]
```diff
@@ -0,0 +1,679 @@
...
+Authors: [Aritra Roy Gosthipaty](https://twitter.com/ariG23498), [Sayak Paul](https://twitter.com/RisingSayak) (equal contribution)
+Date created: 2022/04/12
+Last modified: 2022/04/15
+Description: Probing into various variants of Vision Transformers.
```
Comment (fchollet) [inline examples/vision/probing_vits.py:6]: Avoid "various variants". Let's find a subtitle that adds details compared to the title, e.g. "Looking into the representations learned by different Vision Transformers variants."

**Review by fchollet — COMMENTED** [review]
Thanks for the updates! I applied copyedits.

**Comment by sayakpaul** [comment]
@fchollet, thanks for your help. 

We've applied your suggestions. Let us know if we're good to add the generated files.

**Review by fchollet — COMMENTED** [review]
Thank you!

I edited the title. Please add the generated files! 👍

**Comment by sayakpaul** [comment]
@fchollet ready for you to take over.

**Comment by sayakpaul** [comment]
@fchollet sorry about the reiteration. We're fixing the tables. Generating the files now. Will update here once done.

**Comment by sayakpaul** [comment]
@fchollet things are good to go on our end. We've added all the necessary files.

**Review by fchollet — APPROVED** [review]
Thank you for another great contribution!

**Pull request merged**